astropy/io/fits/_tiled_compression/setup_package.py

```diff
@@ -0,0 +1,41 @@
+# Licensed under a 3-clause BSD style license
+
+import os
+from collections import defaultdict
+
+from setuptools import Extension
+
+from extension_helpers import pkg_config
+
+SRC_DIR = os.path.join(os.path.dirname(__file__), "src")
+
+
+def get_extensions():
+
+    cfg = defaultdict(list)
+    cfg["sources"].extend(
+        [
+            os.path.join(SRC_DIR, "compression.c"),
+            os.path.join(SRC_DIR, "unquantize.c"),
+        ]
+    )
+
+    if int(os.environ.get("ASTROPY_USE_SYSTEM_CFITSIO", 0)) or int(
+        os.environ.get("ASTROPY_USE_SYSTEM_ALL", 0)
+    ):
+        for k, v in pkg_config(["cfitsio"], ["cfitsio"]).items():
+            cfg[k].extend(v)
+    else:
+        cfg["sources"].extend(
+            [
+                os.path.join("cextern", "cfitsio", "lib", "pliocomp.c"),
+                os.path.join("cextern", "cfitsio", "lib", "ricecomp.c"),
+                os.path.join("cextern", "cfitsio", "lib", "fits_hcompress.c"),
+                os.path.join("cextern", "cfitsio", "lib", "fits_hdecompress.c"),
+                os.path.join("cextern", "cfitsio", "lib", "quantize.c"),
+            ]
+        )
+
+    cfg["include_dirs"].append(SRC_DIR)
```

Comment:
This line should be indented I think, i.e. should be done only when not using the system cfitsio.

Reply:
Ah actually not since you need the additional headers defined in astropy. But then some header files (`fitsio2.h` typically) can be found from two different locations when using system astropy. Not sure if that can cause problems, e.g. if system cfitsio's version is different from astropy's one ?

Reply:
(from the sidelines so possibly irrelevant) - I think if astropy provides additional headers, the solution would be to have those in a different file. It may be good to check with @olebole on the system file use.

Reply:
I can also only comment from aside, having checked the few lines above only. My understanding of these lines is that astropy is actually not *providing* the include files in `SRC_DIR`, but *uses* them. In this case, I agree with @saimn that this should only happen when the astropy-provided cfitsio package is used and not a system one.

This is also the case if they are provided to other packages: if astropy was built against a system cfitsio, other packages should refer to the system API, and therefore the system include file if they are going to be linked to the same library as astropy (or otherwise uses it).

Reply:
... but it seems that this PR proposes to use a *different* shared library than cfitsio (and therefor we can drop the "use-system-lib" here). In this case, I think it would be better to use another name (or location) for the include file (if it is provided to others); otherwise there would be an ambiguity and the case "I want to link to both cfitsio *and* astropy's module" would be very difficult.
Having a final location that allows to write
```C
#include "astropy/fitsio2.h"
```
would be already sufficient.

Reply:
Thanks all for the ideas! In fact ``fitsio2.h`` wasn't used (much) by the files in ``astropy/io/fits/_tiled_compression/src`` so I have moved it to the cextern folder and it only gets used if not using the system cfitsio, which should resolve the issue here (see https://github.com/astropy/astropy/pull/14252/commits/e4919d11896450befdfa9423c2fd5ebcb245a8c4)

Reply:
@olebole - just to clarify:

> but it seems that this PR proposes to use a different shared library than cfitsio

this isn't quite right - we are still using cftisio, just a much smaller fraction of it.

Reply:
@saimn - and just going back to your original comment for this discussion, it makes sense to now keep our include dir for both the case when using our bundled CFITSIO and the system install because the header files now no longer duplicate anything in the CFITSIO library.

Reply:
Sorry I wanted to make another comment yesterday but did not have time to do so. 
So actually I'm wondering if allowing to use the system cfitsio still makes sense. Mostly because we are using custom headers to export some functions that cfitsio does not export. So there is no guarantee that it will work with other versions (past or future) of cfitsio, in a sense these functions are internal and their API might change ?

Reply:
In my opinion, it doesn't make sense to keep using system cfitsio here. There is no need to do this overhead for just a few lines of code.

Reply:
I don't fully understand what is considered public/private in a C library so happy to defer to you both. It might be easier to do remove in a follow-up PR as I need to update the docs etc.